specs/3/README.md

```diff
@@ -0,0 +1,272 @@
+
+#### Semaphore Proof Example
+
+```bash
+{
+  merkleTreeDepth: 10,
+  merkleTreeRoot: '4990292586352433503726012711155167179034286198473030768981544541070532815155',
+  nullifier: '17540473064543782218297133630279824063352907908315494138425986188962403570231',
+  message: '32745724963520510550185023804391900974863477733501474067656557556163468591104',
+  scope: '37717653415819232215590989865455204849443869931268328771929128739472152723456',
+  points: [
+    '21668337069844646813015291115284438234607322052337623326830707330064154913250',
+    '5484905467799538881631237123282286864306155680753671338313686933143657835972',
+    '16129789229127169079253218689550197285028424883172925653046098078118792423164',
+    '20777706122379854993524659601832014684665489694335277047215897593373874956681',
+    '6697558559751679943942291885282718275907555268106795371542167431979105110434',
+    '19709269142703129641057076037387702381970592578248722843989118216760760132874',
+    '17493422037248079872314969622558990504818232868144223100447800353776555945950',
+    '20398320346518400096197920333973312490517624241764728676355205902471625807914'
+  ]
+}
+```
+
+### Proof Verification
+
+Proof verification works both off-chain and on-chain (on [EVM-compatible chains](#evm-compatible-chains) where the Groth16 Semaphore verifier can be deployed).
+
+#### Off-chain
+
+Snarkjs is used to verify the proof.
+
+```ts
+return groth16.verify(
+  verificationKey,
+  [merkleTreeRoot, nullifier, hash(message), hash(scope)],
+  unpackGroth16Proof(points)
+)
+```
+
+To learn more about the proof verification in TypeScript/JavaScript, see [the proof verification code](https://github.com/semaphore-protocol/semaphore/blob/main/packages/proof/src/verify-proof.ts).
+
+#### On-chain
+
+Snarkjs is used to generate a Solidity verifier per tree depth. Then, all verifiers are merged into a single one. To know more about the verifier see the [Semaphore Verifier](https://github.com/semaphore-protocol/semaphore/blob/main/packages/contracts/contracts/base/SemaphoreVerifier.sol).
+
+```solidity
+return
+    verifier.verifyProof(
+        [proof.points[0], proof.points[1]],
+        [[proof.points[2], proof.points[3]], [proof.points[4], proof.points[5]]],
+        [proof.points[6], proof.points[7]],
+        [proof.merkleTreeRoot, proof.nullifier, _hash(proof.message), _hash(proof.scope)],
+        proof.merkleTreeDepth
+    );
+```
+
+### Implementation Notes
+
+- Identity generation SHOULD be performed off-chain, in the user's device.
+- Semaphore proof generation SHOULD be performed off-chain, in the user's device.
+
+### Limitations of the implementation
+
+- **Scalability**: As the number of members grows, generating a client-side Merkle proof is no longer feasible. It’s necessary to have a server to do it. This not only is time and data consuming, but could also allow the server to deanonymize the proofs.
```

Comment:
Is this the only way to do it? E.g. more efficient MT could help maybe. Also some rough data on when this becomes a problem would be useful